### DISCUSSION TITLE
Immutability for Store vs EntityStore

### DISCUSSION THREAD

**Discussion by basstune** [salesforce/akita#742] [Q&A]
Hi,

I've been trying Akita out and I have a question about the implementation of `Store` vs `EntityStore` operations. I know immutability is advised to be handled by the developer, but the `EntityStore` operations (add, update, etc.) seem to 'freeze' the object passed as state themselves. This results in an immutable object, which when I retrieve from the store, if I want to mutate it, I have to 'unfreeze it' somehow. I guess this is the intent.

In contrast, I was not able to achieve this using the regular `Store`. However I try to update the store (with a callback or with a plain object), I end up with an object that, when I get back from the store, can mutate and the changes are reflected in the store value.

It seems that immutability comes out of the box for `EntityStore`, but not for `Store`. Wanted to ask if my understanding is correct.

**Comment by NetanelBasal** [comment]
Both will freeze the state. https://github.com/datorama/akita/blob/master/libs/akita/src/lib/store.ts#L181

**Reply by basstune** [reply]
Here is a small demo of creating a `Store` and updating the state directly through the reference I got from `store.geteValue()`:

https://stackblitz.com/edit/angular-ivy-4dafwh?devtoolsheight=33&file=src/app/app.component.ts

**Comment by NetanelBasal** [comment]
I see the issue. We only freeze it if we get a function:

https://github.com/datorama/akita/blob/baa69c9dd5abe916bc9efd13bb36a0404680d7dc/libs/akita/src/lib/store.ts#L181

**Reply by basstune** [reply]
That's right, but, when going through the `update` method of the `Store`, the value always gets resolved (even if coming from a function) before it reaches the `_setState()`

https://github.com/datorama/akita/blob/baa69c9dd5abe916bc9efd13bb36a0404680d7dc/libs/akita/src/lib/store.ts#L251

Is the workaround to use `_setState()` instead of `update()`? Or should I use a `producerFn`?

**Reply by NetanelBasal** [reply]
No, we need to fix it, so that it will work with both.
